src/main/java/org/openstreetmap/atlas/checks/validation/intersections/BoundaryIntersectionCheck.java

```diff
@@ -0,0 +1,403 @@
+            final Set<String> instructions, final Set<AtlasObject> objectsToFlag,
+            final Set<String> matchedTags, final BoundaryPart currentBoundaryPart,
+            final Iterable<LineItem> lineItemsIntersecting, final Set<String> currentMatchedTags)
+    {
+        lineItemsIntersecting.forEach(lineItem ->
+        {
+            final Set<Relation> matchingBoundaries = this.getBoundaries(lineItem).stream()
+                    .filter(boundary -> relationBoundary.getTagToRelation()
+                            .containsKey(boundary.getTag(BOUNDARY).get()))
+                    .filter(boundary -> !relationBoundary
+                            .containsRelationId(boundary.getOsmIdentifier()))
+                    .collect(Collectors.toSet());
+            if (!matchingBoundaries.isEmpty())
+            {
+                currentMatchedTags.addAll(matchingBoundaries.stream()
+                        .map(relation -> relation.getTag(BOUNDARY)).filter(Optional::isPresent)
+                        .map(Optional::get).collect(Collectors.toSet()));
+                objectsToFlag.addAll(matchingBoundaries);
+
+                final List<LineItem> lineItems = this.getLineItems(lineItem);
+                lineItems.forEach(line ->
+                {
+                    objectsToFlag.add(line);
+                    final Coordinate[] intersectingPoints = this.getIntersectionPoints(
+                            currentBoundaryPart.getWktGeometry(), line.toWkt());
+                    final String firstBoundaries = this.objectsToString(
+                            relationBoundary.getRelationsByBoundaryTags(currentMatchedTags));
+                    final String secondBoundaries = this.objectsToString(matchingBoundaries);
+                    if (intersectingPoints.length != 0
+                            && firstBoundaries.hashCode() < secondBoundaries.hashCode())
+                    {
+                        this.addInstruction(instructions, currentBoundaryPart,
+                                line.getOsmIdentifier(), intersectingPoints, firstBoundaries,
+                                secondBoundaries);
+                    }
+                });
+            }
+            matchedTags.addAll(currentMatchedTags);
+        });
+    }
+
+    private Optional<CheckFlag> processRelation(final AtlasObject object)
+    {
+        final Map<String, Relation> tagToRelation = this.getRelationMap(object);
+        final RelationBoundary relationBoundary = new RelationBoundary(tagToRelation,
+                this.getBoundaryParts((Relation) object));
+        final Set<String> instructions = new HashSet<>();
+        final Set<AtlasObject> objectsToFlag = new HashSet<>();
+        final Set<String> matchedTags = new HashSet<>();
+        for (final BoundaryPart currentBoundaryPart : relationBoundary.getBoundaryParts())
+        {
+            final Iterable<LineItem> lineItemsIntersecting = object.getAtlas()
```

Comment:
If the atlas (`object.getAtlas()`) is saved beforehand, it need not be fetched for each BoundaryPart.